server/src/main/java/org/elasticsearch/cluster/metadata/OperatorMetadata.java

```diff
@@ -0,0 +1,311 @@
+         * operator state processing.
+         *
+         * @param errorMetadata {@link OperatorErrorMetadata}
+         * @return {@link Builder}
+         */
+        public Builder errorMetadata(OperatorErrorMetadata errorMetadata) {
+            this.errorMetadata = errorMetadata;
+            return this;
+        }
+
+        /**
+         * Adds an {@link OperatorHandlerMetadata} for this {@link OperatorMetadata}. The handler
+         * metadata is stored in a map, keyed off the {@link OperatorHandlerMetadata} name. Previously
+         * stored {@link OperatorHandlerMetadata} for a given name is overwritten.
+         *
+         * @param handler {@link OperatorHandlerMetadata}
+         * @return {@link Builder}
+         */
+        public Builder putHandler(OperatorHandlerMetadata handler) {
+            this.handlers.put(handler.name(), handler);
+            return this;
+        }
+
+        /**
+         * Builds an {@link OperatorMetadata} from this builder.
+         * @return {@link OperatorMetadata}
+         */
+        public OperatorMetadata build() {
+            return new OperatorMetadata(namespace, version, Collections.unmodifiableMap(handlers), errorMetadata);
+        }
+
+        /**
+         * Serializes the operator metadata to xContent
+         *
+         * @param operatorMetadata {@link OperatorMetadata}
+         * @param builder {@link XContentBuilder}
+         * @param params {@link org.elasticsearch.xcontent.ToXContent.Params}
+         */
+        public static void toXContent(OperatorMetadata operatorMetadata, XContentBuilder builder, ToXContent.Params params)
+            throws IOException {
+            builder.startObject(operatorMetadata.namespace());
+            builder.field(VERSION, operatorMetadata.version());
+            builder.startObject(HANDLERS);
+            var sortedKeys = new TreeSet<>(operatorMetadata.handlers.keySet());
+            for (var key : sortedKeys) {
+                OperatorHandlerMetadata.Builder.toXContent(operatorMetadata.handlers.get(key), builder);
+            }
+            builder.endObject();
+            builder.field(ERRORS_METADATA, operatorMetadata.errorMetadata);
+            builder.endObject();
+        }
+
+        /**
+         * Reads the operator metadata from xContent
+         *
+         * @param parser {@link XContentParser}
+         * @param namespace the operator metadata namespace. See {@link Metadata} operatorMetadata.
+         * @return {@link OperatorMetadata}
+         * @throws IOException
+         */
+        public static OperatorMetadata fromXContent(XContentParser parser, String namespace) throws IOException {
+            OperatorMetadata.Builder builder = new OperatorMetadata.Builder(namespace);
+
+            String currentFieldName = skipNamespace(parser);
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
Could a ConstructingObjectParser be used instead of this loop (same goes for other fromXContent in this PR)? If using that, would a builder even be necessary?

Reply:
I'll look into that, it would be great if we can.